src/nccl_ofi_rdma.c

```diff
@@ -4939,6 +5283,125 @@ static inline int fini_bounce_buffers(nccl_net_ofi_rdma_ep_t *ep)
 	return ret;
 }
 
+static int get_mr_key(nccl_net_ofi_device_t *base_dev, void *mhandle,
+		      uint64_t *mr_key)
+{
+	int ret = 0;
+	nccl_net_ofi_rdma_mr_handle_t *mr_handle = (nccl_net_ofi_rdma_mr_handle_t *)mhandle;
+
+	uint64_t key = fi_mr_key(mr_handle->mr[0]);
+	if (OFI_UNLIKELY(key == FI_KEY_NOTAVAIL)) {
+		ret = -ENOENT;
+		NCCL_OFI_WARN("Error retrieving MR key, leaking key");
+	} else {
+		*mr_key = key;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief	Write using DMA writemsg
+ */
+static int rma_write_impl(nccl_net_ofi_send_comm_t *send_comm, void* src, size_t size, void* desc,
+			  uint64_t dest, uint64_t mr_key, uint64_t flags, nccl_net_ofi_req_t ** base_req)
+{
+	int ret = 0;
+	nccl_net_ofi_rdma_send_comm_t *s_comm = (nccl_net_ofi_rdma_send_comm_t *)send_comm;
+	nccl_net_ofi_rdma_req_t *req = NULL;
+
+	assert(s_comm != NULL);
+
+	/* Support only NCCL_OFI_MAX_REQUESTS inflight requests. */
+	if (OFI_UNLIKELY(s_comm->num_inflight_reqs == NCCL_OFI_MAX_SEND_REQUESTS)) {
+		ret = -EINVAL;
+		NCCL_OFI_WARN("Can not support more than %d inflight requests",
+			      NCCL_OFI_MAX_SEND_REQUESTS);
+		goto error;
+	}
+
+	nccl_net_ofi_rdma_ep_t *ep = (nccl_net_ofi_rdma_ep_t *)s_comm->base.base.ep;
+	assert(ep != NULL);
+
+	ret = process_cq_if_pending(ep);
+	if (ret == -EAGAIN) {
+		/* Network is still busy. Return NULL to NCCL. */
+		*base_req = NULL;
+		ret = 0;
+		goto error;
+	} else if (ret != 0) {
+		goto error;
+	}
+
+	ret = alloc_rdma_write_req(s_comm, ep, src, size, desc, dest, mr_key, flags, &req);
+	if (OFI_UNLIKELY(ret != 0)) {
+		goto error;
+	}
+
+	/*
+	 * At this point, we've successfully inserted a new request,
+	 * so update the num inflight
+	 */
+	(s_comm->num_inflight_reqs)++;
+
+	NCCL_OFI_TRACE_WRITE(req, base_req);
+
+	/* Try posting RMA write with write_inline interface */
+
+	ret = send_progress(req);
+	if (ret == -FI_EAGAIN) {
+		/* Add to pending reqs queue */
+		ret = nccl_ofi_deque_insert_back(ep->pending_reqs_queue, &req->pending_reqs_elem);
+		if (OFI_UNLIKELY(ret != 0)) {
+			NCCL_OFI_WARN("Failed to nccl_ofi_deque_insert_back: %d", ret);
+			goto error;
+		}
+		NCCL_OFI_TRACE_PENDING_INSERT(req);
+	} else if (OFI_UNLIKELY(ret != 0)) {
+		ret = -ENOTSUP;
+		goto error;
+	}
+
+	/* Return request to NCCL */
+	*base_req = &req->base;
+
+	goto exit;
+
+ error:
+	if (req)
+		req->free(req, false);
+	*base_req = NULL;
+ exit:
+	return ret;
+}
+
+/**
+ * @brief	Implementation of iwrite interface. This "interface function" is called, indirectly, from
+ *       	the application
+ */
+
+static int rma_write(nccl_net_ofi_send_comm_t *send_comm, void* src, size_t size, void* mhandle,
+		     uint64_t dest, uint64_t mr_key, nccl_net_ofi_req_t ** base_req)
+{
+	nccl_net_ofi_rdma_mr_handle_t *mr_handle = (nccl_net_ofi_rdma_mr_handle_t *)mhandle;
+	struct fid_mr *rail_mr_handle = mr_handle->mr[0];
+	void *desc = fi_mr_desc(rail_mr_handle);
+	uint64_t flags = 0;
```

Comment:
Why have the `flags` variable in `rma_write()`, but just pass `FI_INJECT` directly in `rma_write_inline()`?

Reply:
Code readability. It gives the reader instant knowledge what the hard-coded value `0` means. When we just use `0` in the function call, you need to know the function signature to know. 

Though, I conduct from you question that you prefer calling the function with value `0`. let me change it.

Reply:
No, I prefer the two functions look the same. Either style is ok with me.

Reply:
Use extra variable `flags` in both functions.